docs/DesignDocs/CustomAttestation_V3.md

```diff
@@ -0,0 +1,423 @@
+
+```C
+   // Old
+   oe_register_attester(oe_sgx_attesterX(), NULL, 0);
+   oe_register_attester(oe_sgx_attesterY(), NULL, 0);
+   oe_register_attester(oe_sgx_attesterZ(), NULL, 0);
+
+   // New
+   oe_sgx_get_attesters(&attesters, &attesters_length);
+   for (n = 0; n < attesters_length; n++)
+      oe_register_attester(&attesters[n], NULL, 0);
+```
+
+2. The attester application can rely on the SDK to select a matching evidence format.
+
+```C
+   // Old
+   const oe_uuid_t attester_formats[3] = {{ATTESTERX_UUID}, {ATTESTERY_UUID}, {ATTESTERZ_UUID}};
+   selected_format = NULL;
+   for (n = 0; n < requested_formats_length && !selected_format; n++)
+   {
+       for (m = 0; m < countof(attester_formats) && !selected_format; m++)
+       {
+           if (requested_formats[n] == attester_formats[m])
+               selected_format = &attester_formats[m];
+       }
+   }
+
+   // New
+   oe_select_attester_format(
+       requested_formats,
+       requested_formats_length,
+       &selected_format);
+```
+
+3. The application gets optional evidence parameters from verifiers and stays TEE-agnostic when generating evidences.
+
+```C
+   // Old
+   uint8_t opt_paramsX[] = { OPT_PARAMS_X };
+   uint8_t opt_paramsY[] = { OPT_PARAMS_Y };
+   if (format == ATTESTERX_UUID)
+   {
+      oe_get_evidence(&format, 0, NULL, 0, opt_paramsX, sizeof(opt_paramsX), &evidence, &evidence_size, NULL, NULL);
+   }
+   else if (format == ATTESTERY_UUID)
+   {
+       oe_get_evidence(&format, 0, NULL, 0, opt_paramsY, sizeof(opt_paramsY), &evidence, &evidence_size, NULL, NULL);
+   }
+
+   // New
+   uint8_t* opt_params;
+   size_t opt_params_size;
+   oe_get_evidence_parameters(&format, &opt_params, &opt_params_size);
+
+   oe_get_evidence(&format, 0, NULL, 0, opt_params, opt_params_size, &evidence, &evidence_size, NULL, NULL);
+```
+
+4. SGX report (referred as local attestation in Open Enclave) and SGX quote (referred as remote attestation in Open Enclave) generation takes different inputs and generates data of different structures. The local/remote attestation concept is SGX specific. To prevent confusion at evidence generation, the OE_EVIDENCE_FLAGS_LOCAL_ATTESTATION and OE_EVIDENCE_FLAGS_REMOTE_ATTESTATION flags were deprecated.
+
+```C
+   // Old
+   oe_get_evidence(&format_sgx, OE_EVIDENCE_FLAGS_LOCAL_ATTESTATION, NULL, 0, opt_params, opt_params_size, &evidence, &evidence_size, NULL, NULL);
+   oe_get_evidence(&format_sgx, OE_EVIDENCE_FLAGS_REMOTE_ATTESTATION, NULL, 0, NULL, 0, &evidence, &evidence_size, NULL, NULL);
+   oe_get_evidence(&format_local_only, OE_EVIDENCE_FLAGS_REMOTE_ATTESTATION, NULL, 0, NULL, 0, &evidence, &evidence_size, NULL, NULL); // Error
+   oe_get_evidence(&format_remote_only, OE_EVIDENCE_FLAGS_LOCAL_ATTESTATION, NULL, 0, NULL, 0, &evidence, &evidence_size, NULL, NULL); // Error
+
+   // New
+   oe_get_evidence(&format_sgx_local, 0, NULL, 0, opt_params, opt_params_size, &evidence, &evidence_size, NULL, NULL);
```

Comment:
I don't see how this is incorporated into the new scenarios and it seems like some knowledge of whether local vs. remote attestation still needs to be exchanged by the attester & verifier. Consider:

1. Verifier _V_ expects to verify local attester _L_ and remote attester _R_. 
   1. _V_ has to register both local and remote SGX verifier plugins. 
1. _R_ also needs to attest locally to a different verifier _V2_
   1. _R_ has to register both local and remote SGX attester plugins. 
1.  Per scenario 1, if _V_ asks to verify _R_, it will give its list to _R_
   1. _R_ supports both, so it comes down to the order _V_ provides
   1. If _V_ orders local ID first, _R_ will give a local report and attestation fails even though both parties support the same format.
   1. If _V_ orders remote ID first, _R_ will succeed, but it will also mean that if _L_ supports remote  SGX attestation too (e.g. _L_ is just a different instance of _R_ enclave code running on the same machine as _V_), then it will always use remote attestation and waste cycles. 

Basically, the scenarios don't account for local vs. remote attestation differences and how the OE APIs need to accommodate them. 

As far as I can tell, the design right now is basically that local attestation is always a special case (unless we propose an OE SDK API to infer common TEE root of trust as part of the attestation plugin infrastructure, for example) and the attester/verifier selection scenarios don't apply; TEEs locally attesting must always just know the desired local evidence format a priori, just like they must know that their verifier is implicitly on the same device for this to work. It's a reasonable enough design, but we should probably call it out here since it doesn't dovetail into any of the rest of the user experience in this proposal.

Reply:
It's challenging trying to accommodate local and remote attestation within the design. However, if we collect the least common parameters to all evidence formats and document/demonstrate the diffs, I think it's better than not supporting known attestation formats.

In the case above, protocol can be defined to continue trying the rest of the supported formats when evidence generation/verification fails (in step 5). In the case of SGX local, if 'target_info' is from a remote platform,EREPORT should fail (does it?).

---
In reply to: [393423331](https://github.com/openenclave/openenclave/pull/2621#discussion_r393423331) [](ancestors = 393423331)

Reply:
> However, if we collect the least common parameters to all evidence formats and document/demonstrate the diffs, I think it's better than not supporting known attestation formats.

Can you elaborate on why this is "better"? In what sense? I don't think that the two problems exist on the same axis of good/bad.

To clarify, I'm pointing out that treating local and remote attesters/verifiers as indistinguishable in the get/select attester/verifier APIs requires more guidance on how to use the APIs effectively, and to me that usually indicates underlying issues in the API. At a minimum, it looks like we need to document:

* The design to generalize the attester/verifier interaction across plugins is primarily intended for remote attestation scenarios. It does not distinguish between local/remote attestation attester/verifier plugins, and registering local attestation plugins in scenarios 1 & 2 with remote plugins leads to the problems described in the previous comment. The samples for scenarios 1 & 2 I would expect to be remote attestation samples.

* While this could be potentially applied to local attestation, it less unlikely to be useful, because enclaves on the same device are going to be constrained to the same TEE local attestation. It's better to document that the workflow for local attestation has not changed, other than the API signature change: apps wanting to perform local attestation must still have _a priori_ expectations that the verifier/attester is also local, and I must know how that expectation maps to a given format ID. I would expect there to be a local attestation sample illustrating this distinction.

Also consider the fact that we already have local and remote attestation samples already based on the `oe_get_report()` and `oe_verify_report()` APIs:
* Why haven't the `oe_get_evidence()` and `oe_verify_evidence()` APIs superseded those APIs? i.e. is there a good reason for their continued coexistence?
* What's the expected transition path from those APIs to the pluggable evidence APIs, which still do take the local vs. remote distinction?

Reply:
> It's challenging trying to accommodate local and remote attestation within the design. 

Actually, thinking about this some more, I'm not sure it's all that hard even. Consider:

* Assertion: `oe_get_evidence()` should not support local/remote flags because format IDs will be specific to either local or remote, making the flags redundant.
   * This is the proposed API change and this is a fair assertion, so let's work with this.

* Assertion: an application already knows if it is performing a local or remote attestation. 
    * This is likely to be true since the app must also setup its IPC or web request etc. 
    * In the cases where the channel could be either, (e.g. loopback) you can always default to the safe lowest common denominator and select remote attestation.

* Suggestion: Developers can benefit from an API that group plugins as being for local or remote attestation.
    * Plugins self-categorize in their registration whether they are local or remote. 
    * Consider a modification of Scenario 2: 
        * I author an enclave that can run on multiple TEEs and wants to attest both locally and remotely.
        * For local attestation, any instance of my enclave can only generate evidence for the TEE I'm running on. It would be nice if:
             * Registration of the attesters automatically does the right thing for the TEE the enclave is running on, without the app code having to do its own detection to register the appropriate format IDs. (e.g. my app code lists all the TEE attesters it wants to support then at runtime, if I running on SGX, only the SGX ones get registered). 
            * I can choose to retrieve at runtime either the set of local or remote attesters registered for the TEE I'm on. (i.e. I can filter for the appropriate format ID based on this local/remote criterion)
           * I can then call `oe_get_evidence` with the provided ID or select from the subset (can't simplify this further since it's plugin dependent)
           * There's also some additional design needed around handling IDs that should self report whether additional `verifier_params` are necessary.

This also gives you a clear through line for migration from `oe_get_report()` APIs to `oe_get_evidence()` APIs